### PR TITLE
HasOne / HasMany must respect $localKey parameter

### PR DESCRIPTION
**Pull request by stephandesouza**
Fixes a bunch of issues.

#1817 #1807 #1692 #1691 #1617 #1540 #1467 

I thing there are more related

### CODE REVIEW

**Comment by stephandesouza** [comment]
ping @jenssegers

**Comment by jenssegers** [comment]
Is this a breaking change?

**Comment by stephandesouza** [comment]
No, just a fix. I updated the tests to force the error when filled.

**Comment by stephandesouza** [comment]
@jenssegers ?

**Comment by stephandesouza** [comment]
Will take a look tomorrow

File: src/Jenssegers/Mongodb/Helpers/QueriesRelationships.php [hunk 1]
```diff
@@ -154,7 +154,7 @@ protected function getConstrainedRelatedIds($relations, $operator, $count)
     protected function getRelatedConstraintKey($relation)
```
[the remark this answers is not in the document]
Comment (Smolevich) [reply src/Jenssegers/Mongodb/Helpers/QueriesRelationships.php]: All relations extends https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/Relations/Relation.php

File: tests/RelationsTest.php [hunk 2]
```diff
@@ -61,24 +61,24 @@ public function testBelongsTo(): void
     public function testHasOne(): void
     {
         $user = User::create(['name' => 'John Doe']);
-        Role::create(['type' => 'admin', 'user_id' => $user->_id]);
+        Role::create(['type' => 'admin', 'roleuser_id' => $user->_id]);
```
Comment (Smolevich) [inline tests/RelationsTest.php]: What happens if we leave name `user_id`?

Comment (divine) [reply tests/RelationsTest.php]: Well, if we leave 'user_id' then there is no way to check that **HasOne / HasMany** respects **$localKey** parameter.

Comment (stephandesouza) [reply tests/RelationsTest.php]: I will make a cleaner test, using a different Model to test this.

File: src/Jenssegers/Mongodb/Helpers/QueriesRelationships.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephandesouza) [reply src/Jenssegers/Mongodb/Helpers/QueriesRelationships.php]: Done with that, also updated `has()` to comply with Laravel signature Relation|string.

**Comment by Smolevich** [comment]
@stephandesouza merge actual master

**Comment by divine** [comment]
@Smolevich @jenssegers any chances to merge it now?

**Comment by Smolevich** [comment]
This PR can have break changes, we need test on this case for example from https://github.com/jenssegers/laravel-mongodb/issues/1902

**Pull request merged**